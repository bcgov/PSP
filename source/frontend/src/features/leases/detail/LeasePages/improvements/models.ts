import { Api_PropertyImprovement } from '@/models/api/PropertyImprovement';

export class ILeaseImprovementsForm {
  improvements: ILeaseImprovementForm[] = [];
}

export class ILeaseImprovementForm {
  id?: number;
  leaseId?: number;
  propertyImprovementTypeId: string = '';
  propertyImprovementType: string = '';
  description: string = '';
  structureSize: string = '';
  address: string = '';
  rowVersion?: number;

  public static fromApi(improvement: Api_PropertyImprovement) {
    const form = new ILeaseImprovementForm();
    form.id = improvement.id ?? undefined;
    form.leaseId = improvement.leaseId ?? undefined;
    form.propertyImprovementTypeId = improvement.propertyImprovementTypeCode?.id ?? '';
    form.propertyImprovementType = improvement.propertyImprovementTypeCode?.description ?? '';
    form.description = improvement.improvementDescription ?? '';
    form.structureSize = improvement.structureSize ?? '';
    form.address = improvement.address ?? '';
    form.rowVersion = improvement.rowVersion ?? undefined;
    return form;
  }

  public static toApi(form: ILeaseImprovementForm) {
    const improvement: Api_PropertyImprovement = {
      id: form.id ?? null,
      leaseId: form.leaseId ?? null,
      lease: null,
      propertyImprovementTypeCode: { id: form.propertyImprovementTypeId },
      improvementDescription: form.description,
      structureSize: form.structureSize,
      address: form.address,
      rowVersion: form.rowVersion,
    };
    return improvement;
  }
}
