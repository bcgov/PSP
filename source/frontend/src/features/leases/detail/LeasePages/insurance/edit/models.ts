import { TypeCodeUtils } from '@/interfaces/ITypeCode';
import { ApiGen_Base_CodeType } from '@/models/api/generated/ApiGen_Base_CodeType';
import { ApiGen_Concepts_Insurance } from '@/models/api/generated/ApiGen_Concepts_Insurance';
import { getEmptyBaseAudit } from '@/models/defaultInitializers';
import { ILookupCode } from '@/store/slices/lookupCodes';
import { NumberFieldValue } from '@/typings/NumberFieldValue';
import { isValidIsoDateTime } from '@/utils';
import { numberFieldToRequiredNumber } from '@/utils/formUtils';

export interface IUpdateFormInsurance {
  insurances: FormInsurance[];
  visibleTypes: string[];
}

export class FormInsurance {
  public isShown!: boolean;
  public isNew!: boolean;
  public isInsuranceInPlaceRadio!: string;

  public id: number | null = null;
  public leaseId: NumberFieldValue = '';
  public insuranceType: ApiGen_Base_CodeType<string> | null = null;
  public otherInsuranceType?: string;
  public coverageDescription?: string;
  public coverageLimit?: NumberFieldValue;
  public expiryDate?: string;
  private rowVersion!: number;

  public static createEmpty(typeCode: ILookupCode, leaseId: number): FormInsurance {
    const model = new FormInsurance();
    model.leaseId = leaseId;
    model.insuranceType = TypeCodeUtils.createFromLookup(typeCode);
    model.coverageLimit = '';
    model.isInsuranceInPlaceRadio = 'no';
    model.isNew = true;
    model.isShown = false;
    return model;
  }

  public static createFromModel(baseModel: ApiGen_Concepts_Insurance): FormInsurance {
    const model = new FormInsurance();
    model.id = baseModel.id;
    model.leaseId = baseModel.leaseId;
    model.insuranceType = baseModel.insuranceType;
    model.otherInsuranceType = baseModel.otherInsuranceType ?? undefined;
    model.coverageDescription = baseModel.coverageDescription ?? undefined;
    model.coverageLimit = baseModel.coverageLimit || '';
    model.expiryDate = isValidIsoDateTime(baseModel.expiryDate) ? baseModel.expiryDate : undefined;
    model.isInsuranceInPlaceRadio = baseModel.isInsuranceInPlace === true ? 'yes' : 'no';
    model.isNew = false;
    model.isShown = true;
    model.rowVersion = baseModel.rowVersion ?? 0;
    return model;
  }

  public toInterfaceModel(): ApiGen_Concepts_Insurance {
    return {
      id: this.id ?? null,
      leaseId: numberFieldToRequiredNumber(this.leaseId),
      insuranceType: this.insuranceType,
      otherInsuranceType: this.otherInsuranceType ?? null,
      coverageDescription: this.coverageDescription ?? null,
      coverageLimit: this.coverageLimit === '' ? null : this.coverageLimit ?? null,
      expiryDate: !isValidIsoDateTime(this.expiryDate) ? null : this.expiryDate ?? null,
      isInsuranceInPlace: this.isInsuranceInPlaceRadio === 'yes' ? true : false,
      ...getEmptyBaseAudit(this.rowVersion),
    };
  }

  public isEqual(other: FormInsurance): boolean {
    return (
      this.isInsuranceInPlaceRadio === other.isInsuranceInPlaceRadio &&
      this.id === other.id &&
      this.insuranceType?.id === other.insuranceType?.id &&
      this.otherInsuranceType === other.otherInsuranceType &&
      this.coverageDescription === other.coverageDescription &&
      this.coverageLimit === other.coverageLimit &&
      this.expiryDate === other.expiryDate
    );
  }
}
