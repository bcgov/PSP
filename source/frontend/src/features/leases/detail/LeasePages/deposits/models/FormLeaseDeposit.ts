import { fromContact, IContactSearchResult, toContact } from '@/interfaces';
import { Api_SecurityDeposit } from '@/models/api/SecurityDeposit';
import { NumberFieldValue } from '@/typings/NumberFieldValue';

export class FormLeaseDeposit {
  public id?: number;
  public leaseId?: number;
  public description: string;
  public amountPaid: NumberFieldValue;
  public depositDate: string;
  public depositTypeCode: string;
  public otherTypeDescription: string;
  public contactHolder?: IContactSearchResult;
  public rowVersion: number;

  private constructor() {
    this.description = '';
    this.amountPaid = '';
    this.depositDate = '';
    this.depositTypeCode = '';
    this.otherTypeDescription = '';
    this.rowVersion = 0;
  }

  public static createEmpty(leaseId: number): FormLeaseDeposit {
    const deposit = new FormLeaseDeposit();
    deposit.leaseId = leaseId;
    return deposit;
  }

  public static fromApi(baseModel: Api_SecurityDeposit): FormLeaseDeposit {
    let model = new FormLeaseDeposit();
    model.leaseId = baseModel.leaseId ?? undefined;
    model.id = baseModel.id ?? undefined;
    model.description = baseModel.description;
    model.amountPaid = baseModel.amountPaid;
    model.depositDate = baseModel.depositDateOnly || '';
    model.depositTypeCode = baseModel.depositType.id;
    model.otherTypeDescription = baseModel.otherTypeDescription || '';
    model.contactHolder =
      baseModel.contactHolder !== null ? fromContact(baseModel.contactHolder) : undefined;

    model.rowVersion = baseModel.rowVersion || 0;
    return model;
  }

  public toApi(): Api_SecurityDeposit {
    return {
      id: this.id ?? null,
      leaseId: this.leaseId ?? null,
      description: this.description,
      amountPaid: this.amountPaid === '' ? 0 : this.amountPaid,
      depositDateOnly: this.depositDate,
      depositType: { id: this.depositTypeCode },
      otherTypeDescription: this.otherTypeDescription === '' ? null : this.otherTypeDescription,
      contactHolder: this.contactHolder !== undefined ? toContact(this.contactHolder) : null,
      depositReturns: [],
      rowVersion: this.rowVersion,
    };
  }
}
