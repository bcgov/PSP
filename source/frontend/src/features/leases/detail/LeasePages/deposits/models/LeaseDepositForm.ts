export class LeaseDepositForm {
  public returnNotes: string = '';
}
