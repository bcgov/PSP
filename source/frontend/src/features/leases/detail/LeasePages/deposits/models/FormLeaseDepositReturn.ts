import { fromContact, IContactSearchResult, toContact } from '@/interfaces';
import { ApiGen_Concepts_SecurityDeposit } from '@/models/api/generated/ApiGen_Concepts_SecurityDeposit';
import { ApiGen_Concepts_SecurityDepositReturn } from '@/models/api/generated/ApiGen_Concepts_SecurityDepositReturn';
import { EpochIsoDateTime } from '@/models/api/UtcIsoDateTime';
import { getEmptyBaseAudit } from '@/models/defaultInitializers';
import { NumberFieldValue } from '@/typings/NumberFieldValue';
import { isValidIsoDateTime } from '@/utils';
import { numberFieldToRequiredNumber } from '@/utils/formUtils';

export class FormLeaseDepositReturn {
  public id?: number;
  public parentDepositId: number;
  public depositTypeCode: string;
  public depositTypeDescription: string;
  public parentDepositOtherDescription: string;
  public terminationDate: string;
  public claimsAgainst: NumberFieldValue;
  public returnAmount: NumberFieldValue;
  public interestPaid: NumberFieldValue;
  public returnDate: string;
  public contactHolder?: IContactSearchResult;
  public rowVersion: number;
  public parentDepositAmount: number;

  private constructor() {
    this.parentDepositId = 0;
    this.depositTypeCode = '';
    this.depositTypeDescription = '';
    this.parentDepositOtherDescription = '';
    this.terminationDate = '';
    this.claimsAgainst = '';
    this.returnAmount = '';
    this.interestPaid = '';
    this.returnDate = '';
    this.parentDepositAmount = 0;
    this.rowVersion = 0;
  }

  public static createEmpty(deposit: ApiGen_Concepts_SecurityDeposit): FormLeaseDepositReturn {
    const returnDeposit = new FormLeaseDepositReturn();
    returnDeposit.parentDepositId = deposit.id || 0;
    returnDeposit.depositTypeCode = deposit.depositType?.id ?? '';
    returnDeposit.depositTypeDescription = deposit.depositType?.description || '';
    returnDeposit.parentDepositOtherDescription = deposit.otherTypeDescription || '';
    returnDeposit.parentDepositAmount = deposit.amountPaid;
    return returnDeposit;
  }

  public static fromApi(
    baseModel: ApiGen_Concepts_SecurityDepositReturn,
    parentDeposit: ApiGen_Concepts_SecurityDeposit,
  ): FormLeaseDepositReturn {
    const model = new FormLeaseDepositReturn();

    // Parent fields
    model.depositTypeCode = parentDeposit.depositType?.id ?? '';
    model.depositTypeDescription = parentDeposit.depositType?.description || '';
    model.parentDepositOtherDescription = parentDeposit.otherTypeDescription || '';
    model.parentDepositAmount = parentDeposit.amountPaid;

    model.id = baseModel.id ?? undefined;
    model.parentDepositId = baseModel.parentDepositId;
    model.terminationDate = isValidIsoDateTime(baseModel.terminationDate)
      ? baseModel.terminationDate
      : '';
    model.claimsAgainst = baseModel.claimsAgainst || '';
    model.returnAmount = baseModel.returnAmount || '';
    model.interestPaid = baseModel.interestPaid || '';
    model.returnDate = isValidIsoDateTime(baseModel.returnDate) ? baseModel.returnDate : '';
    model.contactHolder =
      baseModel.contactHolder !== null ? fromContact(baseModel.contactHolder) : undefined;
    model.rowVersion = baseModel.rowVersion || 0;
    return model;
  }

  public toApi(): ApiGen_Concepts_SecurityDepositReturn {
    return {
      id: this.id ?? null,
      parentDepositId: this.parentDepositId,
      terminationDate: isValidIsoDateTime(this.terminationDate)
        ? this.terminationDate
        : EpochIsoDateTime,
      claimsAgainst: numberFieldToRequiredNumber(this.claimsAgainst),
      returnAmount: numberFieldToRequiredNumber(this.returnAmount),
      interestPaid: numberFieldToRequiredNumber(this.interestPaid),
      returnDate: isValidIsoDateTime(this.returnDate) ? this.returnDate : EpochIsoDateTime,
      contactHolder: this.contactHolder !== undefined ? toContact(this.contactHolder) : null,
      depositType: null,
      ...getEmptyBaseAudit(this.rowVersion),
    };
  }
}
