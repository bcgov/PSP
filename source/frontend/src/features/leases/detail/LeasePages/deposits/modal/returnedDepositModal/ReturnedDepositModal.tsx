import { FormikProps } from 'formik/dist/types';
import { useRef } from 'react';

import GenericModal from '@/components/common/GenericModal';
import { isValidId } from '@/utils';

import { FormLeaseDepositReturn } from '../../models/FormLeaseDepositReturn';
import { ReturnDepositForm } from './ReturnDepositForm';

export interface IReturnedDepositModalProps {
  initialValues?: FormLeaseDepositReturn;
  display: boolean;
  onCancel: () => void;
  onSave: (values: FormLeaseDepositReturn) => void;
}

/**
 * Modal displaying form allowing returned deposits modifications. Save button triggers internal formik validation and submit.
 * @param {IReturnedDepositModalProps} props
 */
export const ReturnedDepositModal: React.FunctionComponent<
  React.PropsWithChildren<IReturnedDepositModalProps>
> = ({ initialValues, display, onCancel, onSave }) => {
  const formikRef = useRef<FormikProps<FormLeaseDepositReturn>>(null);
  const modalTitle = !isValidId(initialValues?.id) ? 'Return a Deposit' : 'Edit a Deposit Return';
  return (
    <GenericModal
      variant="info"
      title={modalTitle}
      display={display}
      okButtonText="Save"
      cancelButtonText="Cancel"
      handleCancel={onCancel}
      handleOk={() => {
        formikRef?.current?.submitForm();
      }}
      message={
        initialValues && (
          <ReturnDepositForm formikRef={formikRef} initialValues={initialValues} onSave={onSave} />
        )
      }
    />
  );
};

export default ReturnedDepositModal;
