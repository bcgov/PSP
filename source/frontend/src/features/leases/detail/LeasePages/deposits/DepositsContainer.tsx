import { Formik } from 'formik';
import { noop } from 'lodash';
import { useContext, useEffect, useState } from 'react';

import GenericModal from '@/components/common/GenericModal';
import LoadingBackdrop from '@/components/common/LoadingBackdrop';
import { LeaseStateContext } from '@/features/leases/context/LeaseContext';
import { LeaseFormModel } from '@/features/leases/models';
import { useSecurityDepositRepository } from '@/hooks/repositories/useSecurityDepositRepository';
import { useSecurityDepositReturnRepository } from '@/hooks/repositories/useSecurityDepositReturnRepository';
import { Api_SecurityDeposit, Api_SecurityDepositReturn } from '@/models/api/SecurityDeposit';

import DepositNotes from './components/DepositNotes/DepositNotes';
import DepositsReceivedContainer from './components/DepositsReceivedContainer/DepositsReceivedContainer';
import DepositsReturnedContainer from './components/DepositsReturnedContainer/DepositsReturnedContainer';
import ReceivedDepositModal from './modal/receivedDepositModal/ReceivedDepositModal';
import ReturnedDepositModal from './modal/returnedDepositModal/ReturnedDepositModal';
import { FormLeaseDeposit } from './models/FormLeaseDeposit';
import { FormLeaseDepositReturn } from './models/FormLeaseDepositReturn';
import { LeaseDepositForm } from './models/LeaseDepositForm';
import * as Styled from './styles';

export interface IDepositsContainerProps {
  onSuccess: () => void;
}

export const DepositsContainer: React.FunctionComponent<
  React.PropsWithChildren<IDepositsContainerProps>
> = props => {
  const { lease } = useContext(LeaseStateContext);
  const {
    getSecurityDeposits: {
      execute: getSecurityDeposits,
      loading,
      response: securityDepositsResponse,
    },
    deleteSecurityDeposit: { execute: deleteSecurityDeposit },
    updateSecurityDeposit: { execute: updateSecurityDeposit },
    addSecurityDeposit: { execute: addSecurityDeposit },
    updateSecurityDepositNote: { execute: updateSecurityDepositNote },
  } = useSecurityDepositRepository();

  const {
    updateSecurityDepositReturn: { execute: updateSecurityDepositReturn },
    addSecurityDepositReturn: { execute: addSecurityDepositReturn },
    deleteSecurityDepositReturn: { execute: deleteSecurityDepositReturn },
  } = useSecurityDepositReturnRepository();

  const securityDeposits: Api_SecurityDeposit[] = securityDepositsResponse ?? [];
  useEffect(() => {
    lease?.id && getSecurityDeposits(lease.id);
  }, [lease, getSecurityDeposits]);
  const depositReturns: Api_SecurityDepositReturn[] =
    securityDeposits?.flatMap((x: Api_SecurityDeposit) => x.depositReturns) ?? [];
  const [editNotes, setEditNotes] = useState<boolean>(false);

  const [showDepositEditModal, setShowEditModal] = useState<boolean>(false);
  const [deleteModalWarning, setDeleteModalWarning] = useState<boolean>(false);
  const [deleteReturnModalWarning, setDeleteReturnModalWarning] = useState<boolean>(false);

  const [showReturnEditModal, setShowReturnEditModal] = useState<boolean>(false);

  const [depositToDelete, setDepositToDelete] = useState<FormLeaseDeposit | undefined>(undefined);
  const [editDepositValue, setEditDepositValue] = useState<FormLeaseDeposit>(
    FormLeaseDeposit.createEmpty(lease?.id ?? 0),
  );

  const [depositReturnToDelete, setDepositReturnToDelete] = useState<
    FormLeaseDepositReturn | undefined
  >(undefined);
  const [editReturnValue, setEditReturnValue] = useState<FormLeaseDepositReturn | undefined>(
    undefined,
  );

  const onAddDeposit = () => {
    lease?.id && setEditDepositValue(FormLeaseDeposit.createEmpty(lease?.id));
    setShowEditModal(true);
  };

  const onEditDeposit = (id: number) => {
    var deposit = securityDeposits.find((x: Api_SecurityDeposit) => x.id === id);
    if (deposit) {
      setEditDepositValue(FormLeaseDeposit.fromApi(deposit));
      setShowEditModal(true);
    }
  };

  const onDeleteDeposit = (id: number) => {
    var deposit = securityDeposits.find((x: Api_SecurityDeposit) => x.id === id);
    if (deposit) {
      setDepositToDelete(FormLeaseDeposit.fromApi(deposit));
      setDeleteModalWarning(true);
    }
  };

  const onReturnDeposit = (id: number) => {
    var deposit = securityDeposits.find((x: Api_SecurityDeposit) => x.id === id);
    if (deposit) {
      setEditReturnValue(FormLeaseDepositReturn.createEmpty(deposit));
      setShowReturnEditModal(true);
    }
  };

  const onDeleteDepositConfirmed = async () => {
    if (lease && lease.id && lease.rowVersion && depositToDelete) {
      await deleteSecurityDeposit(lease.id, depositToDelete.toApi());
      setDepositToDelete(undefined);
      setDeleteModalWarning(false);
      getSecurityDeposits(lease.id);
      props.onSuccess();
    }
  };

  const onDeleteDepositReturnConfirmed = async () => {
    if (lease && lease.id && lease.rowVersion && depositReturnToDelete) {
      await deleteSecurityDepositReturn(lease.id, depositReturnToDelete.toApi());
      setDepositReturnToDelete(undefined);
      setDeleteReturnModalWarning(false);
      getSecurityDeposits(lease.id);
      props.onSuccess();
    }
  };

  /**
   * Send the save request (either an update or an add). Use the response to update the parent lease.
   * @param depositForm
   */
  const onSaveDeposit = async (depositForm: FormLeaseDeposit) => {
    if (lease && lease.id) {
      const updatedSecurityDeposit = depositForm.id
        ? await updateSecurityDeposit(lease.id, depositForm.toApi())
        : await addSecurityDeposit(lease.id, depositForm.toApi());
      if (!!updatedSecurityDeposit?.id) {
        setEditDepositValue(FormLeaseDeposit.createEmpty(lease.id));
        setShowEditModal(false);
        getSecurityDeposits(lease.id);
        props.onSuccess();
      }
    } else {
      console.error('Lease information incomplete');
    }
  };

  const onEditReturnDeposit = (id: number) => {
    var deposit = depositReturns.find(x => x.id === id);
    if (deposit) {
      var parentDeposit = securityDeposits.find(x => x.id === deposit?.parentDepositId);
      if (parentDeposit) {
        setEditReturnValue(FormLeaseDepositReturn.fromApi(deposit, parentDeposit));
        setShowReturnEditModal(true);
        lease?.id && getSecurityDeposits(lease.id);
      } else {
        console.error('Parent deposit incomplete');
      }
    }
  };

  const onDeleteDepositReturn = (id: number) => {
    var deposit = depositReturns.find(x => x.id === id);
    if (deposit) {
      var parentDeposit = securityDeposits.find(
        (x: Api_SecurityDeposit) => x.id === deposit?.parentDepositId,
      );
      if (parentDeposit) {
        setDepositReturnToDelete(FormLeaseDepositReturn.fromApi(deposit, parentDeposit));
        setDeleteReturnModalWarning(true);
      } else {
        console.error('Parent deposit incomplete');
      }
    }
  };

  /**
   * Send the save request (either an update or an add). Use the response to update the parent lease.
   * @param returnDepositForm
   */
  const onSaveReturnDeposit = async (returnDepositForm: FormLeaseDepositReturn) => {
    if (lease && lease.id) {
      let request: Api_SecurityDepositReturn = returnDepositForm.toApi();
      const securityDepositReturn = request?.id
        ? await updateSecurityDepositReturn(lease.id, request)
        : await addSecurityDepositReturn(lease.id, request);
      if (!!securityDepositReturn?.id) {
        setDepositReturnToDelete(undefined);
        setShowReturnEditModal(false);
        getSecurityDeposits(lease.id);
        props.onSuccess();
      }
    } else {
      console.error('Lease information incomplete');
    }
  };

  const initialValues = LeaseFormModel.fromApi(lease);

  return (
    <>
      <LoadingBackdrop show={loading} parentScreen />
      <Formik initialValues={{ ...new LeaseDepositForm(), ...initialValues }} onSubmit={noop}>
        {formikProps => (
          <Styled.DepositsContainer>
            <DepositsReceivedContainer
              securityDeposits={securityDeposits}
              onAdd={onAddDeposit}
              onEdit={onEditDeposit}
              onDelete={onDeleteDeposit}
              onReturn={onReturnDeposit}
            />

            <DepositsReturnedContainer
              securityDeposits={securityDeposits}
              depositReturns={depositReturns}
              onEdit={onEditReturnDeposit}
              onDelete={onDeleteDepositReturn}
            />

            <DepositNotes
              disabled={!editNotes}
              onEdit={() => setEditNotes(true)}
              onSave={async (notes: string) => {
                lease?.id && (await updateSecurityDepositNote(lease.id, notes));
                setEditNotes(false);
                props.onSuccess();
              }}
              onCancel={() => {
                setEditNotes(false);
                formikProps.setFieldValue('returnNotes', lease?.returnNotes ?? '');
              }}
            />

            <GenericModal
              variant="warning"
              display={deleteModalWarning}
              title="Delete Deposit"
              message={`Are you sure you want to remove the deposit?`}
              handleOk={() => onDeleteDepositConfirmed()}
              okButtonText="OK"
              closeButton
              setDisplay={setDeleteModalWarning}
            />
            <GenericModal
              variant="warning"
              display={deleteReturnModalWarning}
              title="Delete Deposit Return"
              message={`Are you sure you want to remove this deposit return?`}
              handleOk={() => onDeleteDepositReturnConfirmed()}
              okButtonText="OK"
              closeButton
              setDisplay={setDeleteReturnModalWarning}
            />

            <ReceivedDepositModal
              display={showDepositEditModal}
              initialValues={editDepositValue}
              onCancel={() => {
                lease?.id && setEditDepositValue(FormLeaseDeposit.createEmpty(lease.id));
                setShowEditModal(false);
              }}
              onSave={onSaveDeposit}
            />

            <ReturnedDepositModal
              display={showReturnEditModal}
              initialValues={editReturnValue}
              onCancel={() => {
                setEditReturnValue(undefined);
                setShowReturnEditModal(false);
              }}
              onSave={onSaveReturnDeposit}
            />
          </Styled.DepositsContainer>
        )}
      </Formik>
    </>
  );
};

export default DepositsContainer;
