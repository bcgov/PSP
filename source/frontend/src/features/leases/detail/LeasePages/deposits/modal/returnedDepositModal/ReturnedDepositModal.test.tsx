import userEvent from '@testing-library/user-event';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { createMemoryHistory } from 'history';

import { getMockDepositReturns, getMockDeposits } from '@/mocks/deposits.mock';
import { mockLookups } from '@/mocks/lookups.mock';
import { lookupCodesSlice } from '@/store/slices/lookupCodes';
import { act, fillInput, renderAsync, RenderOptions, waitFor } from '@/utils/test-utils';

import { FormLeaseDepositReturn } from '../../models/FormLeaseDepositReturn';
import ReturnedDepositModal, { IReturnedDepositModalProps } from './ReturnedDepositModal';

const history = createMemoryHistory();
const mockAxios = new MockAdapter(axios);
const onSave = vi.fn();
const onCancel = vi.fn();
const storeState = {
  [lookupCodesSlice.name]: { lookupCodes: mockLookups },
};

describe('ReturnedDepositModal component', () => {
  const setup = async (
    renderOptions: RenderOptions &
      Partial<IReturnedDepositModalProps> & {
        initialValues?: any;
      } = {},
  ) => {
    // render component under test
    const component = await renderAsync(
      <ReturnedDepositModal
        onSave={onSave}
        onCancel={onCancel}
        display={true}
        initialValues={FormLeaseDepositReturn.createEmpty(getMockDeposits()[0])}
      />,
      {
        ...renderOptions,
        history,
        store: storeState,
      },
    );

    return {
      component,
    };
  };

  beforeEach(() => {
    mockAxios.resetHistory();
    onSave.mockReset();
    onCancel.mockReset();
  });
  it('renders as expected', async () => {
    const { component } = await setup({});

    expect(component.asFragment()).toMatchSnapshot();
  });
  it('renders with data as expected', async () => {
    const { component } = await setup({
      initialValues: FormLeaseDepositReturn.createEmpty(getMockDeposits()[0]),
    });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('submits all filled out fields as expected', async () => {
    const {
      component: { getByText },
    } = await setup({
      initialValues: FormLeaseDepositReturn.fromApi(
        getMockDepositReturns()[0],
        getMockDeposits()[0],
      ),
    });

    await fillInput(document.body, 'terminationDate', '2020-01-01', 'datepicker');
    await fillInput(document.body, 'returnDate', '2020-01-02', 'datepicker');
    await fillInput(document.body, 'claimsAgainst', '1,000.00');
    await fillInput(document.body, 'returnAmount', '2,000.00');
    await fillInput(document.body, 'interestPaid', '1.00');
    await fillInput(document.body, 'contactHolder.id', 'p1');

    const saveButton = getByText('Save');
    await act(async () => userEvent.click(saveButton));
    await waitFor(() => expect(onSave).toHaveBeenCalled());
    expect(onSave).toHaveBeenCalledWith({
      claimsAgainst: 1000,
      depositTypeCode: 'PET',
      depositTypeDescription: '',
      id: undefined,
      parentDepositAmount: 500,
      parentDepositId: 1,
      parentDepositOtherDescription: '',
      returnAmount: 2000,
      interestPaid: 1,
      returnDate: '2020-01-02',
      rowVersion: 0,
      terminationDate: '2020-01-01',
      contactHolder: { id: 'p1' },
    });
  });

  it('calls onCancel when cancel button clicked', async () => {
    const {
      component: { getByText },
    } = await setup({});
    const cancelButton = getByText('Cancel');
    userEvent.click(cancelButton);
    expect(onCancel).toHaveBeenCalled();
  });
});
