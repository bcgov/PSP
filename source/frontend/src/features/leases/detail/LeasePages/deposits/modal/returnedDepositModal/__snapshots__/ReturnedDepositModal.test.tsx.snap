// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReturnedDepositModal component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  <div />
</DocumentFragment>
`;

exports[`ReturnedDepositModal component renders with data as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div />
  <div />
</DocumentFragment>
`;
